### PR TITLE
Fixes for changes in location of apple clang's libc++ header locations

### PR DESCRIPTION
**Pull request by gartung**
These are ifdef'ed with __APPLE__ so they should not interfere with linux compilations. I had to add /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1 to ROOT_ICLUDE_PATH before /usr/include/c++/4.2.1 to ensure that the correct header were picked up by genreflex. I also ensure that the cxxflags had this correct as well.

### CODE REVIEW

**Comment by gartung** [comment]
@Dr15Jones

**Comment by Dr15Jones** [comment]
My understanding is `<ext/functional>` is for dealing with proposed library changes. We should be using pure C++14. What was the exact reason for using that header?

File: DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h [hunk 1]
```diff
@@ -16,7 +16,9 @@
 #include "DataFormats/BTauReco/interface/CandIPTagInfo.h"
 #include "DataFormats/Candidate/interface/VertexCompositePtrCandidate.h"
 #include <functional>
+#ifndef __APPLE__
 #include <ext/functional>
+#endif
```
Comment (slava77) [inline DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h]: should there be an "else"?

File: DataFormats/Common/interface/RangeMap.h [hunk 2]
```diff
@@ -77,7 +82,9 @@ namespace edm {
     /// to use only comparators provided with CMSSW release.
     template<typename CMP> 
     range get(ID id, CMP comparator) const {
+#ifndef __APPLE__
       using namespace __gnu_cxx;
```
Comment (Dr15Jones) [inline DataFormats/Common/interface/RangeMap.h]: Is this still needed for gcc?

File: DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h (hunk 1 above)
Comment (gartung) [inline DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h]: There is a #include <functional> above which is apparently all that is needed.

File: DataFormats/Common/interface/RangeMap.h (hunk 2 above)
Comment (gartung) [inline DataFormats/Common/interface/RangeMap.h]: I don't know but the classes.h also has references to __gnu_cxx

input_line_9:66:3: error: use of undeclared identifier '__gnu_cxx'
  __gnu_cxx::new_allocator<char> dummy21;
  ^

File: DataFormats/Common/interface/RangeMap.h [hunk 3]
```diff
@@ -77,7 +82,9 @@ namespace edm {
...
+#endif
```
Comment (slava77) [inline DataFormats/Common/interface/RangeMap.h]: please check if this compiles without ```using namespace __gnu_cxx ```
unless you know offhand that it is needed for something here

File: DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h (hunk 1 above)
Comment (slava77) [inline DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h]: ah. What's in the ``` ext/functional ``` that's not already in ```functional ```, do you happen to know?

Comment (Dr15Jones) [inline DataFormats/BTauReco/interface/TemplatedSecondaryVertexTagInfo.h]: We removed `ext/functional` and the code still compiles. I bet someone included that when we were using C++11 in order to get some future C++14 code. Now that we are using C++14 the use of the header appears to be unnecessary.

File: DataFormats/Common/interface/RangeMap.h (hunk 3 above)
Comment (gartung) [inline DataFormats/Common/interface/RangeMap.h]: It compiles without this.

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/17168	73db55d
- technical change moving to mainstream system includes
- jenkins tests pass

**Comment by davidlange6** [comment]
+1

**Pull request merged**